OC work

wjam opened the pull request:
Highlights of this PR includes the following changes:

* Addition of docker-compose file to make it easier to run Redis Enterprise locally
* Migration to a typed client
* Restructure tests to allow them to be run inside GitHub CI
* Expansion of test coverage

wjam committed:
Add Docker Compose file (#2)

Add a docker-compose.yml file and accompanying shell script to make it easier to start up a test Redis Enterprise database so that the plugin tests can be run against it

wjam committed:
Add small typed client for interacting with the cluster API (#1)

Also:
* start using the client in the plugin
* expand tests to confirm that the user does/doesn't exist
* expand the .gitignore file to cover more files
* add EditorConfig file to help maintain consistent styling
* use hclog rather than the fmt package for logging so that log messages are correctly visible when running the plugin in Vault
* use the Vault SDK `credsutil` package to generate the username

wjam committed:
Add initial pipeline for building & testing the plugin

Add a Makefile to make it easy to build & test the plugin:
* `make` to build the plugin, run the tests, and perform various static analysis
* `make start-docker` and `make stop-docker` to spin up/down a Redis Enterprise within Docker
* `make configure-docker` to configure Vault, running within Docker, to use the plugin to generate credentials within the Redis Enterprise cluster

Add a `.goreleaser.yml` file to control the building of releases when the correct tags have been pushed

Add version information that the released binaries can inject tag and commit information into

trentrosenbaum committed:
Initial restructure for individual tests and constant usernames

trentrosenbaum committed:
Initial restructure for individual tests and constant usernames

wjam committed:
Merge pull request #3 from opencredo/workflow

Add initial pipeline for building & testing the plugin

trentrosenbaum committed:
Updates to introduce go-vcr into test.  Tests are grouped in init, new, update and delete

trentrosenbaum committed:
Updates to introduce go-vcr into test.  Tests are grouped in init, new, update and delete

trentrosenbaum committed:
Updates to expand the recorder

trentrosenbaum committed:
Updates to expand the recorder

trentrosenbaum committed:
Refactors the creation/init of the SimpleRESTClient to allow recording through go-vcr

trentrosenbaum committed:
Refactors the creation/init of the SimpleRESTClient to allow recording through go-vcr

trentrosenbaum committed:
Renames newuser test at the cluster level

trentrosenbaum committed:
Renames newuser test at the cluster level

trentrosenbaum committed:
Updates SimpleRESTCLient init with BaseURL strings.TrimSuffix and explicately sets the enableACL flasg in each test.

trentrosenbaum committed:
Updates SimpleRESTCLient init with BaseURL strings.TrimSuffix and explicately sets the enableACL flasg in each test.

trentrosenbaum committed:
Allows the fixtures to be disbled through RS_DISABLE_FIXTURES env variable

trentrosenbaum committed:
Allows the fixtures to be disbled through RS_DISABLE_FIXTURES env variable

trentrosenbaum committed:
Updated with formatting

trentrosenbaum committed:
Updated with formatting

Commit:
Merge branch 'oc-work' into initial_test_refactor

trentrosenbaum committed:
Merge remote-tracking branch 'origin/initial_test_refactor' into initial_test_refactor

trentrosenbaum committed:
Updated the fixtures to reference localhost instead of 127.0.0.1

trentrosenbaum committed:
Updates the url field with the strings.TrimSuffix(url, /)

wjam committed:
Merge pull request #4 from opencredo/initial_test_refactor

Initial test refactor

wjam committed:
Prefer the use structs for handling config over maps

Replace the use of `map[string]interface{}` with structs to simplify the code and provide compile-time type safety

Also move the relevant code from the root of the repository into internal/, to match the https://github.com/golang-standards/project-layout

Also start getting GitHub to perform CI

Also get goreleaser to produce binaries rather than compressed files, so those who download the plugin can refer to the checksum file for the expected checksum value

wjam committed:
Merge pull request #5 from opencredo/structify

Prefer the use structs for handling config over maps

wjam committed:
Split single file into multiple files

Split out `redisenterprise.go` into separate files to avoid having a single monolithic file

wjam committed:
Merge pull request #6 from opencredo/separate_files

Split single file into multiple files

wjam committed:
Migrate ACL code to use the SDK

Rewrite the functionality to dynamic creation of roles to use the SDK.

Also:
* Harden the ACL code to avoid problems where many role permissions are being added at the same time
* Avoid explicitly removing the role permissions before deleting the generated role as Redis Enterprise will remove the role permissions for you

wjam committed:
Merge pull request #7 from opencredo/acl_sdk

Migrate ACL code to use the SDK

wjam committed:
Add tests to verify failure roll back behaviour

Add tests to ensure that the relevant resources are deleted if there's a failure when creating the new user, and verify that any failure when cleaning up gets wrapped by the original error

Also change to use `errors.Is`

wjam committed:
Merge pull request #9 from opencredo/acl_tests

Add tests to verify failure roll back behaviour

wjam committed:
Add test to verify new user with role and ACL defined

Verify that the plugin will return an error if creating a user with a mismatched role and ACL.

Also verify that the newly created user is assigned to the correct role

wjam committed:
Merge pull request #10 from opencredo/role_acl_test

Add test to verify new user with role and ACL defined

Commit:
Update development documentation (#8)

* Updated README.md to focus on the local build, test and development.
* Introduced the ARCHITECTURE.md to represent a focus for architecture content.

wjam committed:
Add test to verify that an ACL user can be correctly deleted

Verify that an ACL user has their role deleted as well. Also add a make command to easily view the code coverage

wjam committed:
Merge pull request #11 from opencredo/delete_acl_user_test

Add test to verify that an ACL user can be correctly deleted

trentrosenbaum commented:
Just to add to this PR.  The README.md is focused right now on the local testing and executing the acceptance tests against the docker image.  I am going to expand the docs to include standalone Vault and then draw together the k8s steps.  These will come in another PR to follow.

wjam committed:
Remove the need for recorded interactions

Replace the need for prerecorded interactions for CI with the Redis Enterprise Kubernetes operator instead.

Also remove the need for being able to mock out the username generation, have the tests use the 'real' `New` function, and add test to ensure passwords are correctly removed from errors

wjam committed:
Merge pull request #3 from opencredo/k8s-operator-ci

Remove the need for recorded interactions

wjam committed:
Cache dependencies between CI runs

Speed up builds by caching Go dependencies between runs.

Also change CI pipeline to use the status reported by the Kubernetes resources rather than attempting to perform curl commands

wjam committed:
Merge pull request #5 from opencredo/ci-dep-cache

Cache dependencies between CI runs

wjam committed:
Ensure original error gets wrapped

wjam committed:
Merge pull request #6 from opencredo/wrap-err

Ensure original error gets wrapped

wjam committed:
Add a link to the custom DB engine

wjam committed:
Changes from code review

wjam committed:
Merge pull request #7 from opencredo/db-engine-link

Add a link to the custom DB engine
